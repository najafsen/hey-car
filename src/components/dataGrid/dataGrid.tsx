import styled from "styled-components";
import { mdMediaQuery } from "../../constants/styles";

export const color = `#acacac`;
export const borderColor = `#e4e4e4`;
export const highlightColor = `#f5f5f5`;

const Table = styled.div.attrs({
  role: "table",
})`
  border: 1px solid #e4e4e4;
`;

interface RowProps {
  columnCount: number;
}

const Row = styled.div.attrs({
  role: "row",
})`
  display: grid;
  grid-auto-flow: row;

  :nth-child(2n) {
    background-color: ${highlightColor};
  }

  :not(:last-child) {
    border-bottom: 1px solid ${borderColor};
  }

  ${mdMediaQuery} {
    grid-auto-flow: column;
    grid-template-columns: ${(p: RowProps) =>
      `repeat(${p.columnCount}, ${100 / p.columnCount}%)`};
  }
`;

const Cell = styled.div.attrs({
  role: "cell",
})`
  color: ${color};
  padding: 12px;

  :not(:last-child) {
    border-bottom: 1px dotted ${borderColor};
  }

  ${mdMediaQuery} {
    :not(:last-child) {
      border-bottom: none;
    }
  }
`;

interface DataGridProps<TItem> {
  columns: (keyof TItem)[];
  items: TItem[];
}

/**
 * Very basic data grid!
 */
export const DataGrid = <TItem extends unknown>({
  columns,
  items,
}: DataGridProps<TItem>) => (
  <Table>
    {items.map((item, index) => (
      <Row key={index} columnCount={columns.length}>
        {columns.map((column) => (
          <Cell key={column.toString()}>{item[column] as any}</Cell>
        ))}
      </Row>
    ))}
  </Table>
);
