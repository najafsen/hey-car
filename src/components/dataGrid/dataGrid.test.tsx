import { render } from "@testing-library/react";
import { DataGrid } from "./dataGrid";

interface TestDataGridInterface {
  name: string;
  id: string;
}

const items: TestDataGridInterface[] = [
  { name: "name1", id: "id1" },
  { name: "name2", id: "id2" },
  { name: "name3", id: "id3" },
  { name: "name4", id: "id4" },
  { name: "name5", id: "id5" },
  { name: "name6", id: "id6" },
];
describe("DataGrid", () => {
  it("should render correctly", () => {
    const { asFragment } = render(
      <DataGrid<TestDataGridInterface> columns={["name", "id"]} items={items} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should have 'name6id6' in 6th row", () => {
    const { getAllByRole } = render(
      <DataGrid<TestDataGridInterface> columns={["name", "id"]} items={items} />
    );

    expect(getAllByRole("row")[5].textContent).toBe("name6id6");
  });
});
