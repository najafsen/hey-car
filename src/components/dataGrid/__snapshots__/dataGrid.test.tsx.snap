// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataGrid should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM fPfiLM"
    role="table"
  >
    <div
      class="sc-gtsrHT Emqft"
      role="row"
    >
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        name1
      </div>
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        id1
      </div>
    </div>
    <div
      class="sc-gtsrHT Emqft"
      role="row"
    >
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        name2
      </div>
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        id2
      </div>
    </div>
    <div
      class="sc-gtsrHT Emqft"
      role="row"
    >
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        name3
      </div>
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        id3
      </div>
    </div>
    <div
      class="sc-gtsrHT Emqft"
      role="row"
    >
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        name4
      </div>
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        id4
      </div>
    </div>
    <div
      class="sc-gtsrHT Emqft"
      role="row"
    >
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        name5
      </div>
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        id5
      </div>
    </div>
    <div
      class="sc-gtsrHT Emqft"
      role="row"
    >
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        name6
      </div>
      <div
        class="sc-dlnjwi cJyYRj"
        role="cell"
      >
        id6
      </div>
    </div>
  </div>
</DocumentFragment>
`;
